extras/cdrom_id/cdrom_id.8
.TH CDROM_ID 8 "January 2009" "" "Linux Administrator's Manual"
.SH NAME
cdrom_id \- udev callout to determine the capabilities of optical drives and media
.SH SYNOPSIS
.B cdrom_id
[\fB--export\fP] [\fB--debug\fP] \fIdevice\fP
.br
.B cdrom_id
\fB--help\fP
.SH "DESCRIPTION"
.B cdrom_id
is normally called from a udev rule, to provide udev with the list of
capabilities of a optical drive.
.SH USAGE
.B cdrom_id
opens the device node specified at the commandline and prints the
discovered capabilities.
.SH OPTIONS
The following commandline switches are supported to specify what cdrom_id
should print:
.TP
.B -x, --export
Print all values as environment keys. This is the default at present, so this
option is currently redundant.
.TP
.B -d, --debug
Print a debug trace.
.TP
.B -h, --help
Print usage help.
.RE
.SH SEE ALSO
.BR udev (7)
.SH AUTHORS
Developed by Kay Sievers <kay.sievers@vrfy.org>.



NEWS
udev 141
========
Bugfixes.

udev 140
========
Bugfixes.

"udevadm settle" now optionally accepts a range of events to wait for,
instead of waiting for "all" events.

udev 139
========
Bugfixes.

The installed watch for block device metadata changes is now removed
during event hadling, because some (broken) tools may be called from udev
rules and (wrongly) open the device with write access. After the finished
event handling the watch is restored.

udev 138
========
Bugfixes.

Device nodes can be watched for changes with inotify with OPTIONS="watch".
If closed after being opened for writing, a "change" uevent will occur.
/dev/disk/by-{label,uuid}/* symlinks will be automatically updated.

udev 137
========
Bugfixes.

The udevadm test command has no longer a --force option, nodes and symlinks
are always updated with a test run now.

The udevd daemon can be started with --resolve-names=never to avoid all user
and group lookups (e.g. in cut-down systems) or --resolve-names=late to
lookup user and groups every time events are handled.

udev 136
========
Bugfixes.

We are currently merging the Ubuntu rules in the udev default rules,
and get one step closer to provide a common Linux /dev setup, regarding
device names, symlinks, and default device permissions. On udev startup,
we now expect the following groups to be resolvable to their ids with
glibc's getgrnam():
  disk, cdrom, floppy, tape, audio, video, lp, tty, dialout, kmem.
LDAP setups need to make sure, that these groups are always resolvable at
bootup, with only the rootfs mounted, and without network access available.

Some systems may need to add some new, currently not used groups, or need
to add some users to new groups, but the cost of this change is minimal,
compared to the pain the current, rather random, differences between the
various distributions cause for upstream projects and third-party vendors.

In general, "normal" users who log into a machine should never be a member
of any such group, but the device-access should be managed by dynamic ACLs,
which get added and removed for the specific users on login/logout and
session activity/inactivity. These groups are only provided for custom setups,
and mainly system services, to allow proper privilege separation.
A video-streaming daemon uid would be a member of "audio" and "video", to get
access to the sound and video devices, but no "normal" user should ever belong
to the "audio" group, because he could listen to the built-in microphone with
any ssh-session established from the other side of the world.

/dev/serial/by-{id,path}/ now contains links for ttyUSB devices,
which do not depend on the kernel device name. As usual, unique
devices - only a single one per product connected, or a real
USB serial number in the device - are always found with the same
name in the by-id/ directory.
Completely identical devices may overwrite their names in by-id/
and can only be found reliably in the by-path/ directory. Devices
specified by by-path/ must not change their connection, like the
USB port number they are plugged in, to keep their name.

To support some advanced features, Linux 2.6.22 is the oldest supported
version now. The kernel config with enabled SYSFS_DEPRECATED is no longer
supported. Older kernels should still work, and devices nodes should be
reliably created, but some rules and libudev will not work correctly because
the old kernels do not provide the expected information or interfaces.

udev 135
========
Bugfixes.

Fix for a possible segfault while swapping network interface names in udev
versions 131-134.

udev 134
========
Bugfixes.

The group "video" is part of the default rules now.

udev 133
========
Bugfix for kernels using SYSFS_DEPRECATED* option and finding parent
block devices in some cases. No common distro uses this option anymore,
and we do not get enough testing for this and recent udev versions. If
this option is not needed to run some old distro with a new kernel,
it should be disabled in the kernel config.

Bugfix for the $links substitution variable, which may crash if no links
are created. This should not happen in usual setups because we always
create /dev/{block,char}/ links.

The strings of the parsed rules, which are kept in memory, no longer
contain duplicate entries, or duplicate tails of strings. This, and the
new rules parsing/matching code reduces the total in-memory size of
a huge distro rule sets to 0.08 MB, compared to the 1.2MB of udev
version 130.

The export of DEVTYPE=disk/partition got removed from the default
rules. This value is available from the kernel. The pnp shell script
modprobe hack is removed from the default rules. ACPI devices have _proper_
modalias support and take care of the same functionality.
Installations which support old kernels, but install current default
udev rules may want to add that to the compat rules file.

Libvolume_id now always probes for all known filesystems, and does not
stop at the first match. Some filesystems are marked as "exclusive probe",
and if any other filesytem type matches at the same time, libvolume_id
will, by default, not return any probing result. This is intended to prevent
mis-detection with conflicting left-over signatures found from earlier
file system formats. That way, we no longer depend on the probe-order
in case of multiple competing signatures. In some setups the kernel allows
to mount a volume with just the old filesystem signature still in place.
This may damage the new filesystem and cause data-loss, just by mounting
it. Because volume_id can not decide which one the correct signature is,
the wrong signatures need to be removed manually from the volume, or the
volume needs to be reformatted, to enable filesystem detection and possible
auto-mounting.

udev 132
========
Fix segfault if compiled without optimization and dbg() does not get
compiled out and uses variables which are not available.

udev 131
========
Bugfixes. (And maybe new bugs. :))


extras/cdrom_id/cdrom_id.c
/*
 * cdrom_id - optical drive and media information prober
 *
 * Copyright (C) 2008 Kay Sievers <kay.sievers@vrfy.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE 1
#endif

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <limits.h>
#include <fcntl.h>
#include <errno.h>
#include <getopt.h>
#include <scsi/sg.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <linux/cdrom.h>

#include "../../udev/udev.h"

static int debug;

static void log_fn(struct udev *udev, int priority,
		   const char *file, int line, const char *fn,
		   const char *format, va_list args)
{
	if (debug) {
		fprintf(stderr, "%s: ", fn);
		vfprintf(stderr, format, args);
	} else {
		vsyslog(priority, format, args);
	}
}

/* device info */
static unsigned int cd_cd_rom;
static unsigned int cd_cd_r;
static unsigned int cd_cd_rw;
static unsigned int cd_dvd_rom;
static unsigned int cd_dvd_r;
static unsigned int cd_dvd_rw;
static unsigned int cd_dvd_ram;
static unsigned int cd_dvd_plus_r;
static unsigned int cd_dvd_plus_rw;
static unsigned int cd_dvd_plus_r_dl;
static unsigned int cd_dvd_plus_rw_dl;
static unsigned int cd_bd;
static unsigned int cd_bd_r;
static unsigned int cd_bd_re;
static unsigned int cd_hddvd;
static unsigned int cd_hddvd_r;
static unsigned int cd_hddvd_rw;
static unsigned int cd_mo;
static unsigned int cd_mrw;
static unsigned int cd_mrw_w;

/* media info */
static unsigned int cd_media_cd_rom;
static unsigned int cd_media_cd_r;
static unsigned int cd_media_cd_rw;
static unsigned int cd_media_dvd_rom;
static unsigned int cd_media_dvd_r;
static unsigned int cd_media_dvd_rw;
static unsigned int cd_media_dvd_ram;
static unsigned int cd_media_dvd_plus_r;
static unsigned int cd_media_dvd_plus_rw;
static unsigned int cd_media_dvd_plus_r_dl;
static unsigned int cd_media_dvd_plus_rw_dl;
static unsigned int cd_media_bd;
static unsigned int cd_media_bd_r;
static unsigned int cd_media_bd_re;
static unsigned int cd_media_hddvd;
static unsigned int cd_media_hddvd_r;
static unsigned int cd_media_hddvd_rw;
static unsigned int cd_media_mo;
static unsigned int cd_media_mrw;
static unsigned int cd_media_mrw_w;

static const char *cd_media_state;
static unsigned int cd_media_session_next;
static unsigned int cd_media_session_count;
static unsigned int cd_media_track_count;
static unsigned int cd_media_track_count_data;
static unsigned int cd_media_track_count_audio;
static unsigned long long int cd_media_session_last_offset;

#define ERRCODE(s)	((((s)[2] & 0x0F) << 16) | ((s)[12] << 8) | ((s)[13]))
#define SK(errcode)	(((errcode) >> 16) & 0xF)
#define ASC(errcode)	(((errcode) >> 8) & 0xFF)
#define ASCQ(errcode)	((errcode) & 0xFF)

static void info_scsi_cmd_err(struct udev *udev, char *cmd, int err)
{
	if (err == -1) {
		info(udev, "%s failed\n", cmd);
		return;
	}
	info(udev, "%s failed with SK=%Xh/ASC=%02Xh/ACQ=%02Xh\n", cmd, SK(err), ASC(err), ASCQ(err));
}

struct scsi_cmd {
	struct cdrom_generic_command cgc;
	union {
		struct request_sense s;
		unsigned char u[18];
	} _sense;
	struct sg_io_hdr sg_io;
};

static void scsi_cmd_set(struct udev *udev, struct scsi_cmd *cmd, size_t i, int arg)
{
	if (i == 0) {
		memset(cmd, 0x00, sizeof(struct scsi_cmd));
		cmd->cgc.quiet = 1;
		cmd->cgc.sense = &cmd->_sense.s;
		memset(&cmd->sg_io, 0, sizeof(cmd->sg_io));
		cmd->sg_io.interface_id = 'S';
		cmd->sg_io.mx_sb_len = sizeof(cmd->_sense);
		cmd->sg_io.cmdp = cmd->cgc.cmd;
		cmd->sg_io.sbp = cmd->_sense.u;
		cmd->sg_io.flags = SG_FLAG_LUN_INHIBIT | SG_FLAG_DIRECT_IO;
	}
	cmd->sg_io.cmd_len = i + 1;
	cmd->cgc.cmd[i] = arg;
}

#define CHECK_CONDITION 0x01

static int scsi_cmd_run(struct udev *udev, struct scsi_cmd *cmd, int fd, unsigned char *buf, size_t bufsize)
{
	int ret = 0;

	cmd->sg_io.dxferp = buf;
	cmd->sg_io.dxfer_len = bufsize;
	cmd->sg_io.dxfer_direction = SG_DXFER_FROM_DEV;
	if (ioctl(fd, SG_IO, &cmd->sg_io))
		return -1;

	if ((cmd->sg_io.info & SG_INFO_OK_MASK) != SG_INFO_OK) {
		errno = EIO;
		ret = -1;
		if (cmd->sg_io.masked_status & CHECK_CONDITION) {
			ret = ERRCODE(cmd->_sense.u);
			if (ret == 0)
				ret = -1;
		}
	}
	return ret;
}

static int cd_capability_compat(struct udev *udev, int fd)
{
	int capabilty;

	capabilty = ioctl(fd, CDROM_GET_CAPABILITY, NULL);
	if (capabilty < 0) {
		info(udev, "CDROM_GET_CAPABILITY failed\n");
		return -1;
	}

	if (capabilty & CDC_CD_R)
		cd_cd_r = 1;
	if (capabilty & CDC_CD_RW)
		cd_cd_rw = 1;
	if (capabilty & CDC_DVD)
		cd_dvd_rom = 1;
	if (capabilty & CDC_DVD_R)
		cd_dvd_r = 1;
	if (capabilty & CDC_DVD_RAM)
		cd_dvd_ram = 1;
	if (capabilty & CDC_MRW)
		cd_mrw = 1;
	if (capabilty & CDC_MRW_W)
		cd_mrw_w = 1;
	return 0;
}

static int cd_inquiry(struct udev *udev, int fd) {
	struct scsi_cmd sc;
	unsigned char inq[128];
	int err;

	scsi_cmd_set(udev, &sc, 0, 0x12);
	scsi_cmd_set(udev, &sc, 4, 36);
	scsi_cmd_set(udev, &sc, 5, 0);
	err = scsi_cmd_run(udev, &sc, fd, inq, 36);
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "INQUIRY", err);
		return -1;
	}

	if ((inq[0] & 0x1F) != 5) {
		info(udev, "not an MMC unit\n");
		return -1;
	}

	info(udev, "INQUIRY: [%.8s][%.16s][%.4s]\n", inq + 8, inq + 16, inq + 32);
	return 0;
}

static int cd_profiles(struct udev *udev, int fd)
{
	struct scsi_cmd sc;
	unsigned char header[8];
	unsigned char profiles[512];
	unsigned int cur_profile;
	unsigned int len;
	unsigned int i;
	int err;

	scsi_cmd_set(udev, &sc, 0, 0x46);
	scsi_cmd_set(udev, &sc, 1, 0);
	scsi_cmd_set(udev, &sc, 8, sizeof(header));
	scsi_cmd_set(udev, &sc, 9, 0);
	err = scsi_cmd_run(udev, &sc, fd, header, sizeof(header));
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "GET CONFIGURATION", err);
		return -1;
	}

	len = 4 + (header[0] << 24 | header[1] << 16 | header[2] << 8 | header[3]);
	info(udev, "GET CONFIGURATION: number of profiles %i\n", len);
	if (len > sizeof(profiles)) {
		info(udev, "invalid number of profiles\n");
		return -1;
	}

	scsi_cmd_set(udev, &sc, 0, 0x46);
	scsi_cmd_set(udev, &sc, 1, 1);
	scsi_cmd_set(udev, &sc, 6, len >> 16);
	scsi_cmd_set(udev, &sc, 7, len >> 8);
	scsi_cmd_set(udev, &sc, 8, len);
	scsi_cmd_set(udev, &sc, 9, 0);
	err = scsi_cmd_run(udev, &sc, fd, profiles, len);
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "GET CONFIGURATION", err);
		return -1;
	}

	/* device profiles */
	for (i = 12; i < profiles[11]; i += 4) {
		unsigned int profile = (profiles[i] << 8 | profiles[i + 1]);
		if (profile == 0)
			continue;
		info(udev, "profile 0x%02x\n", profile);

		switch (profile) {
		case 0x03:
		case 0x04:
		case 0x05:
			cd_mo = 1;
		break;
		case 0x10:
			cd_dvd_rom = 1;
			break;
		case 0x12:
			cd_dvd_ram = 1;
			break;
		case 0x13:
		case 0x14:
			cd_dvd_rw = 1;
			break;
		case 0x1B:
			cd_dvd_plus_r = 1;
			break;
		case 0x1A:
			cd_dvd_plus_rw = 1;
			break;
		case 0x2A:
			cd_dvd_plus_rw_dl = 1;
			break;
		case 0x2B:
			cd_dvd_plus_r_dl = 1;
			break;
		case 0x40:
			cd_bd = 1;
			break;
		case 0x41:
		case 0x42:
			cd_bd_r = 1;
			break;
		case 0x43:
			cd_bd_re = 1;
			break;
		case 0x50:
			cd_hddvd = 1;
			break;
		case 0x51:
			cd_hddvd_r = 1;
			break;
		case 0x52:
			cd_hddvd_rw = 1;
			break;
		default:
			break;
		}
	}

	/* current media profile */
	cur_profile = header[6] << 8 | header[7];
	info(udev, "current profile 0x%02x\n", cur_profile);
	if (cur_profile == 0) {
		info(udev, "no current profile, assuming no media\n");
		return -1;
	}

	switch (cur_profile) {
	case 0x03:
	case 0x04:
	case 0x05:
		cd_media_mo = 1;
		break;
	case 0x08:
		cd_media_cd_rom = 1;
		break;
	case 0x09:
		cd_media_cd_r = 1;
		break;
	case 0x0a:
		cd_media_cd_rw = 1;
		break;
	case 0x10:
		cd_media_dvd_rom = 1;
		break;
	case 0x11:
		cd_media_dvd_r = 1;
		break;
	case 0x12:
		cd_media_dvd_ram = 1;
		break;
	case 0x13:
	case 0x14:
		cd_media_dvd_rw = 1;
		break;
	case 0x1B:
		cd_media_dvd_plus_r = 1;
		break;
	case 0x1A:
		cd_media_dvd_plus_rw = 1;
		break;
	case 0x2A:
		cd_media_dvd_plus_rw_dl = 1;
		break;
	case 0x2B:
		cd_media_dvd_plus_r_dl = 1;
		break;
	case 0x40:
		cd_media_bd = 1;
		break;
	case 0x41:
	case 0x42:
		cd_media_bd_r = 1;
		break;
	case 0x43:
		cd_media_bd_re = 1;
		break;
	case 0x50:
		cd_media_hddvd = 1;
		break;
	case 0x51:
		cd_media_hddvd_r = 1;
		break;
	case 0x52:
		cd_media_hddvd_rw = 1;
		break;
	default:
		break;
	}
	return 0;
}

static int cd_media_info(struct udev *udev, int fd)
{
	struct scsi_cmd sc;
	unsigned char header[32];
	static const char *media_status[] = {
		"blank",
		"appendable",
		"complete",
		"other"
	};
	int err;

	scsi_cmd_set(udev, &sc, 0, 0x51);
	scsi_cmd_set(udev, &sc, 8, sizeof(header));
	scsi_cmd_set(udev, &sc, 9, 0);
	err = scsi_cmd_run(udev, &sc, fd, header, sizeof(header));
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "READ DISC INFORMATION", err);
		return -1;
	};

	info(udev, "disk type %02x\n", header[8]);

	if ((header[2] & 3) < 4)
		cd_media_state = media_status[header[2] & 3];
	if ((header[2] & 3) != 2)
		cd_media_session_next = header[10] << 8 | header[5];
	cd_media_session_count = header[9] << 8 | header[4];
	cd_media_track_count = header[11] << 8 | header[6];

	return 0;
}

static int cd_media_toc(struct udev *udev, int fd)
{
	struct scsi_cmd sc;
	unsigned char header[12];
	unsigned char toc[2048];
	unsigned int len, i;
	unsigned char *p;
	int err;

	scsi_cmd_set(udev, &sc, 0, 0x43);
	scsi_cmd_set(udev, &sc, 6, 1);
	scsi_cmd_set(udev, &sc, 8, sizeof(header));
	scsi_cmd_set(udev, &sc, 9, 0);
	err = scsi_cmd_run(udev, &sc, fd, header, sizeof(header));
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "READ TOC", err);
		return -1;
	}

	len = (header[0] << 8 | header[1]) + 2;
	info(udev, "READ TOC: len: %d\n", len);
	if (len > sizeof(toc))
		return -1;
	if (len < 2)
		return -1;

	/* empty media has no tracks */
	if (len < 8)
		return 0;

	scsi_cmd_set(udev, &sc, 0, 0x43);
	scsi_cmd_set(udev, &sc, 6, header[2]); /* First Track/Session Number */
	scsi_cmd_set(udev, &sc, 7, len >> 8);
	scsi_cmd_set(udev, &sc, 8, len);
	scsi_cmd_set(udev, &sc, 9, 0);
	err = scsi_cmd_run(udev, &sc, fd, toc, len);
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "READ TOC (tracks)", err);
		return -1;
	}

	for (p = toc+4, i = 4; i < len-8; i += 8, p += 8) {
		unsigned int block;
		unsigned int is_data_track;

		is_data_track = (p[1] & 0x04) != 0;

		block = p[4] << 24 | p[5] << 16 | p[6] << 8 | p[7];
		info(udev, "track=%u info=0x%x(%s) start_block=%u\n",
		     p[2], p[1] & 0x0f, is_data_track ? "data":"audio", block);

		if (is_data_track)
			cd_media_track_count_data++;
		else
			cd_media_track_count_audio++;
	}

	scsi_cmd_set(udev, &sc, 0, 0x43);
	scsi_cmd_set(udev, &sc, 2, 1); /* Session Info */
	scsi_cmd_set(udev, &sc, 8, 12);
	scsi_cmd_set(udev, &sc, 9, 0);
	err = scsi_cmd_run(udev, &sc, fd, header, sizeof(header));
	if ((err < 0)) {
		info_scsi_cmd_err(udev, "READ TOC (multi session)", err);
		return -1;
	}
	len = header[4+4] << 24 | header[4+5] << 16 | header[4+6] << 8 | header[4+7];
	info(udev, "last track %u starts at block %u\n", header[4+2], len);
	cd_media_session_last_offset = (unsigned long long int)len * 2048;
	return 0;
}

int main(int argc, char *argv[])
{
	struct udev *udev;
	static const struct option options[] = {
		{ "export", no_argument, NULL, 'x' },
		{ "debug", no_argument, NULL, 'd' },
		{ "help", no_argument, NULL, 'h' },
		{}
	};
	const char *node = NULL;
	int export = 0;
	int fd = -1;
	int rc = 0;

	udev = udev_new();
	if (udev == NULL)
		goto exit;

	logging_init("cdrom_id");
	udev_set_log_fn(udev, log_fn);

	while (1) {
		int option;

		option = getopt_long(argc, argv, "dxh", options, NULL);
		if (option == -1)
			break;

		switch (option) {
		case 'd':
			debug = 1;
			if (udev_get_log_priority(udev) < LOG_INFO)
				udev_set_log_priority(udev, LOG_INFO);
			break;
		case 'x':
			export = 1;
			break;
		case 'h':
			printf("Usage: cdrom_id [options] <device>\n"
			       "  --export        export key/value pairs\n"
			       "  --debug         debug to stderr\n"
			       "  --help          print this help text\n\n");
			goto exit;
		default:
			rc = 1;
			goto exit;
		}
	}

	node = argv[optind];
	if (!node) {
		err(udev, "no device\n");
		fprintf(stderr, "no device\n");
		rc = 1;
		goto exit;
	}

	fd = open(node, O_RDONLY | O_NONBLOCK);
	if (fd < 0) {
		info(udev, "unable to open '%s'\n", node);
		rc = 1;
		goto exit;
	}
	info(udev, "probing: '%s'\n", node);

	/* same data as original cdrom_id */
	if (cd_capability_compat(udev, fd) < 0) {
		rc = 1;
		goto exit;
	}

	/* check drive */
	if (cd_inquiry(udev, fd) < 0) {
		rc = 2;
		goto exit;
	}

	/* read drive and possibly current profile */
	if (cd_profiles(udev, fd) < 0)
		goto print;

	/* get session/track info */
	if (cd_media_toc(udev, fd) < 0)
		goto print;

	/* get writable media state */
	if (cd_media_info(udev, fd) < 0)
		goto print;

print:
	printf("ID_CDROM=1\n");
	if (cd_cd_rom)
		printf("ID_CDROM_CD=1\n");
	if (cd_cd_r)
		printf("ID_CDROM_CD_R=1\n");
	if (cd_cd_rw)
		printf("ID_CDROM_CD_RW=1\n");
	if (cd_dvd_rom)
		printf("ID_CDROM_DVD=1\n");
	if (cd_dvd_r)
		printf("ID_CDROM_DVD_R=1\n");
	if (cd_dvd_rw)
		printf("ID_CDROM_DVD_RW=1\n");
	if (cd_dvd_ram)
		printf("ID_CDROM_DVD_RAM=1\n");
	if (cd_dvd_plus_r)
		printf("ID_CDROM_DVD_PLUS_R=1\n");
	if (cd_dvd_plus_rw)
		printf("ID_CDROM_DVD_PLUS_RW=1\n");
	if (cd_dvd_plus_r_dl)
		printf("ID_CDROM_DVD_PLUS_R_DL=1\n");
	if (cd_dvd_plus_rw_dl)
		printf("ID_CDROM_DVD_PLUS_RW_DL=1\n");
	if (cd_bd)
		printf("ID_CDROM_BD=1\n");
	if (cd_bd_r)
		printf("ID_CDROM_BD_R=1\n");
	if (cd_bd_re)
		printf("ID_CDROM_BD_RE=1\n");
	if (cd_hddvd)
		printf("ID_CDROM_HDDVD=1\n");
	if (cd_hddvd_r)
		printf("ID_CDROM_HDDVD_R=1\n");
	if (cd_hddvd_rw)
		printf("ID_CDROM_HDDVD_RW=1\n");
	if (cd_mo)
		printf("ID_CDROM_MO=1\n");
	if (cd_mrw)
		printf("ID_CDROM_MRW=1\n");
	if (cd_mrw_w)
		printf("ID_CDROM_MRW_W=1\n");

	if (cd_media_mo)
		printf("ID_CDROM_MEDIA_MO=1\n");
	if (cd_media_mrw)
		printf("ID_CDROM_MEDIA_MRW=1\n");
	if (cd_media_mrw_w)
		printf("ID_CDROM_MEDIA_MRW_W=1\n");
	if (cd_media_cd_rom)
		printf("ID_CDROM_MEDIA_CD=1\n");
	if (cd_media_cd_r)
		printf("ID_CDROM_MEDIA_CD_R=1\n");
	if (cd_media_cd_rw)
		printf("ID_CDROM_MEDIA_CD_RW=1\n");
	if (cd_media_dvd_rom)
		printf("ID_CDROM_MEDIA_DVD=1\n");
	if (cd_media_dvd_r)
		printf("ID_CDROM_MEDIA_DVD_R=1\n");
	if (cd_media_dvd_ram)
		printf("ID_CDROM_MEDIA_DVD_RAM=1\n");
	if (cd_media_dvd_rw)
		printf("ID_CDROM_MEDIA_DVD_RW=1\n");
	if (cd_media_dvd_plus_r)
		printf("ID_CDROM_MEDIA_DVD_PLUS_R=1\n");
	if (cd_media_dvd_plus_rw)
		printf("ID_CDROM_MEDIA_DVD_PLUS_RW=1\n");
	if (cd_media_dvd_plus_rw_dl)
		printf("ID_CDROM_MEDIA_DVD_PLUS_RW_DL=1\n");
	if (cd_media_dvd_plus_r_dl)
		printf("ID_CDROM_MEDIA_DVD_PLUS_R_DL=1\n");
	if (cd_media_bd)
		printf("ID_CDROM_MEDIA_BD=1\n");
	if (cd_media_bd_r)
		printf("ID_CDROM_MEDIA_BD_R=1\n");
	if (cd_media_bd_re)
		printf("ID_CDROM_MEDIA_BD_RE=1\n");
	if (cd_media_hddvd)
		printf("ID_CDROM_MEDIA_HDDVD=1\n");
	if (cd_media_hddvd_r)
		printf("ID_CDROM_MEDIA_HDDVD_R=1\n");
	if (cd_media_hddvd_rw)
		printf("ID_CDROM_MEDIA_HDDVD_RW=1\n");

	if (cd_media_state != NULL)
		printf("ID_CDROM_MEDIA_STATE=%s\n", cd_media_state);
	if (cd_media_session_next > 0)
		printf("ID_CDROM_MEDIA_SESSION_NEXT=%d\n", cd_media_session_next);
	if (cd_media_session_count > 0)
		printf("ID_CDROM_MEDIA_SESSION_COUNT=%d\n", cd_media_session_count);
	if (cd_media_track_count > 0)
		printf("ID_CDROM_MEDIA_TRACK_COUNT=%d\n", cd_media_track_count);
	if (cd_media_track_count_audio > 0)
		printf("ID_CDROM_MEDIA_TRACK_COUNT_AUDIO=%d\n", cd_media_track_count_audio);
	if (cd_media_track_count_data > 0)
		printf("ID_CDROM_MEDIA_TRACK_COUNT_DATA=%d\n", cd_media_track_count_data);
	if (cd_media_session_last_offset > 0)
		printf("ID_CDROM_MEDIA_SESSION_LAST_OFFSET=%llu\n", cd_media_session_last_offset);
exit:
	if (fd >= 0)
		close(fd);
	udev_unref(udev);
	logging_close();
	return rc;
}

